### ISSUE TITLE
High-DPI Support

### ISSUE THREAD

**Issue by vkbo** [vkbo/novelWriter#280]
There are a few issues with novelWriter on high DPI screens. A number of margins and default sizes have hard coded values that are too small on scaled displays, and especially the QSwitch widget doesn't scale at all.

Text currently scales fine, so do most icons.

**Comment by johnblommers** [comment]
I did a bit of testing of my own. I'm running Ubuntu 19.10 with the MATE desktop environment. My  monitor will do 3840x2160@30Hz and 1920x1080@120Hz so I tested with these resolutions. However, despite setting the HiDPI setting to automagic the system does not recalculate the DPI and leaves it at 96DPI when I change the screen size. So I manually set the DPI to 192 for the 3840x2160 screen size and to 96 DPI for the 1920x1080 screen size.

For each screen size and DPI setting I launched novelWriter open to the sample document, pressed F11 to go full screen, and took a screenshot <kbd>Prt Sc</kbd>. Then I opened each screenshot in its own layer with the GIMP, 192DPI on top, set it to 50% transparency, and scaled it down 50% to match the size of the 96dpi screenshot in the layer below.

As you've noted, novelWriter does a pretty good job of maintaining its proportions between these two display sizes. IMHO the differences are not significant to the user.

Attached are to two screenshots plus the GIMP two-layer file.

**96dpi* screenshot:

![96dpi](https://user-images.githubusercontent.com/8484427/83927537-1037e480-a742-11ea-80d4-63920fce9110.png)


**192dpi** screenshot

![192dpi](https://user-images.githubusercontent.com/8484427/83927569-2a71c280-a742-11ea-9e2a-9ded90af9cb2.png)


**GIMP two-layer file zip**

[192dpi-96dpi-layers.xcf.zip](https://github.com/vkbo/novelWriter/files/4738768/192dpi-96dpi-layers.xcf.zip)

**Comment by vkbo** [comment]
The big question here is what branch you downloaded or checked out. The fix is in `dev` at the moment, not in `master`.

Qt handles scaling fairly well on its own, especially the font sizing. A lot of GUI elements in novelWriter derive their sizes from the font size. However, a bunch of fixed sizes exist, and all user settings, both those in preferences, and those saved from resizing widows and columns, are saved explicitly in pixels.

The latter doesn't translate if the app is opened alternately on low and high DPI screens. I have been testing this myself on my work laptop which has a 4K screen @ 192 DPI. It is normally connected to a large desktop screen at @ 96 DPI. Switching between those two messes up a bunch of things. For instance, as can be seen from your screenshots, the column widths in the trees and tables, the document margin, and if you'd used a fixed document width, that too. It also affects grid layout spacing, like the details panel below the tree, and the width of text entry boxes on the various dialogs.

Now, in the `dev` branch after merging #285, user defined sizes, also widths and heights from dragging things, is saved to the config file after being divided by the scale factor, and applied after being multiplied with it. That means that the apparent sizes shouldn't change whether you open on a high or a low DPI screen.

What it still does not support, is a mich of high and low DPI screens. It only detects the scale of your primary screen, and sets the sizes on launch.

**Comment by johnblommers** [comment]
**Comments**:

1. I always test the novelWriter master branch unless asked otherwise. that's what I did here.
1. Your multi-DPI testing setup with dual monitors at different DPI is the harshest environment for this kind of work.
1. From my reading and experience, different Linux distros, versions and desktop environments have spotty and differing ways to support screen DPI or HiDPI as they say now. I understand the objective is to be able to run the same application on different DPI screens and still have the application window sized the same number of inches. 
1. I find that neither Ubuntu 16.04 LTS running Unity, nor Ubuntu 19.10 running MATE change the DPI setting  away from its default 96 dpi.
1. I did not appreciate what a mess HiDPI would be!

**Question**: In your testing, which Linux distro automatically sets the DPI setting properly?  I'm finding this: 

```bash
xdpyinfo | grep dot
  resolution:    96x96 dots per inch
```

**Comment by vkbo** [comment]
The Qt library has a way to extract the DPI, and I'm not sure where it gets it from, or if it;s consistent with what other tools report. In any case, if you run with `--verbose`, it should pre printed to the log output during initialisation of the Theme class, which is before the GUI is drawn.

Anyway, I've now merged the dev branch to master, so feel free to check it out. When I add new features I try to do it separately from the master branch in case I need to add a fix to the latest release, but don't want to pull inn experimental stuff. But the latest features seem to work fine, so I merged them in. I will pin the 0.8 release with these new features in a week or so.

As it stands now, novelWriter is feature complete for the set of initial features I wanted to have. There are of course others that we've been discussing, online as well as offline with my former colleague, but this is on a longer time schedule.

Right now, the aim is to polish what we've got, and ensure it's stable. Then I want to wind down development a bit, and do some actual writing!
